File: services/engine/engine/stress.py
"""Stress controls applied on top of any decision.

- `collection_delay_days` shifts every movement exposed to collection delay.
- `main_customer_lost` removes the main customer's share from exposed receipts.
- `capital_injection_cents` adds a day-0 inflow to cash.
"""

from __future__ import annotations

from dataclasses import dataclass, replace

from engine.cash.movements import Movement
from engine.cash.profile import CompanyProfile
from engine.errors import ParameterError


@dataclass(frozen=True)
class Stress:
    collection_delay_days: int = 0
    main_customer_lost: bool = False
    capital_injection_cents: int = 0

    def validate(self) -> None:
        if self.collection_delay_days < 0 or self.capital_injection_cents < 0:
            raise ParameterError("stress controls must be non-negative")

    def apply(self, movements: list[Movement], profile: CompanyProfile) -> list[Movement]:
        self.validate()
        out: list[Movement] = []
        share = profile.main_customer_concentration or 0.0
        for m in movements:
            if self.main_customer_lost and m.exposure == "main_customer":
                kept = int(round(m.amount_cents * (1 - share)))
                m = replace(m, amount_cents=kept)
            out.append(m)
        if self.capital_injection_cents:
            out.append(
                Movement("stress.capital_injection", "cash", "in", 0, self.capital_injection_cents)
            )
        return out

File: services/domain/internal/scenario/domain/stress.go
package domain

import "fmt"

// Stress holds the three controls applied on top of any decision (PRD §3.2,
// block 3). They are part of the parameterization, so they are part of the
// cache key and are echoed back in the state document.
type Stress struct {
	CollectionDelayDays   int
	MainCustomerLost      bool
	CapitalInjectionCents int64
}

// Validate enforces engine/stress.py:Stress.validate.
func (s Stress) Validate() error {
	if s.CollectionDelayDays < 0 || s.CapitalInjectionCents < 0 {
		return fmt.Errorf("stress controls must be non-negative")
	}
	return nil
}
